### DISCUSSION TITLE
Possibility for native http streams?

### DISCUSSION THREAD

**Discussion by haderp** [gtxaspec/wz_mini_hacks#243] [Q&A]
Would it be possible to get a native mjpeg or html5 streaming server included with the firmware?  My main goal is being able to use the cameras directly in my 3D printer software, but it only supports mjpeg or webrtc streams.  I've setup a server to convert the rtsp streams to webrtc, but wanted to know if the possibility of a native server for mjpeg or webrtc could be made.  I apologize in advance if I failed to adequately RTFM and there already exists a method to do this.

**Comment by virmaior** [comment]
This is the same issue as https://github.com/gtxaspec/wz_mini_hacks/discussions/197

Unfortunately what we can get native from the camera doesn't have webrtc built-in. Can run ffmepg to convert. I believe it can even be run on the camera itself.

**Reply by haderp** [reply]
That's what I have been finding with my limited experience with ffmpeg.  Since I don't need the audio portion of the stream and in all honesty the resolution isn't even a big factor for what I hope to achieve I'll delve further in on the ffmpeg route.  One thing I did test was a binary built from the openipc-firmware called [mJpegStreamer](https://github.com/openipcamera/openipc-firmware/blob/master/firmware_mod/v2/bin/mJpegStreamer).  If i were using v2 hardware I think it would have run without issues since it will run, but doesn't find the expected video.  Thanks for taking the time to reply, if nothing else it tells me I am on the right path, and just need to keep throwing noodles at the wall until one sticks.

**Comment by virmaior** [comment]
what about https://github.com/valbok/mjpeg-over-http ?

**Reply by haderp** [reply]
That is ultimately the avenue I'll probably take for simplicity and for resource conservation.  I hope to have  some time to setup the build environment and make it.  For now I've got it being served out through docker containers for each camera which is of course a pig, but it gets the job done.  Thanks for all the input, I appreciate your time.

**Reply by virmaior** [reply]
I'd actually like to solve the same problem myself.

`v4l2rtspserver` itself actually can offer HLS but when I adde the parameter manually I get:

```
MPEG2TransportStreamFromESSource: input buffer too small; increase "MPEG2TransportStreamFromESSource::maxInputESFrameSize" by at least 49744 bytes!
AddH26xMarkerFilter::afterGettingFrame(): buffer too small truncated:129648 bufferSize:329648
MPEG2TransportStreamFromESSource: input buffer too small; increase "MPEG2TransportStreamFromESSource::maxInputESFrameSize" by at least 129648 bytes!
Segmentation fault (core dumped)
```

If I can get that solved, then the app is dead to me, because I can then look at all channels through a browser.

**Reply by haderp** [reply]
I think I know your solution to this one.  It's being caused by the max input frame size being set to 100000 bytes in MPEG2TransportStreamFromESSource.cpp and the input frame is bigger than that max.  Increasing that value to accommodate, I assume, a 1920 x 1080  and rebuilding I suspect would resolve this.  Further reference here:

[](https://github.com/mpromonet/v4l2rtspserver/issues/206)

Hope that helps get you to your goal.  Cheers

**Reply by virmaior** [reply]
reference is clobbered but that's very helpful as to why it's crashing. For my own application the low res image is probably fine so I should turn on the second channel to get it.

**Reply by haderp** [reply]
Sorry about that, here is the link:  https://github.com/mpromonet/v4l2rtspserver/issues/206  (You'd think some people would use the preview or proofread before replying.)

Something to note is that it will still segfault depending on, I assume, the initial frame size that happens to be grabbed when ffmpeg starts to encode the output.

**Reply by gtxaspec** [reply]
i'll look into this.

**Comment by endertable** [comment]
Hi @gtxaspec just a question, how hard would this be for that mjpeg-over-http to compile or add on?

**Comment by gtxaspec** [comment]
working on compiling these

**Comment by vincom** [comment]
im not the fluent in streaming protocols.....but i did get streaming via port 80.
on the web page for the cam, webserver running of course, right click the stream to copy the image url.
u just need an app to get it streaming and not a static jpg. in chrome theres an extension "myipcam" that can use the jpg snapshot path and used blue iris to stream using said path. im assuming any app or program, could not get it to work on vlc but im a newb with vlc.

**Reply by haderp** [reply]
That is definitely a good and viable option for creating a stream based on current mechanisms in place.  I did something similar to that idea.  I created a sub directory under www then copied the elements from the car directory into the new directory.  Then I stripped out everything except the jpeg elements, modified the new stream.js to call the backend cgi script every 150ms and managed to have a fairly fluid and stable stream.  I've tested it for the past couple of days and haven't found that it causes any issues, however I always have a certain amount of reservation when using something in a way the developer may not have intended because of potential issues it could cause.  Var a more elegant solution that pulls from the hardware directly it will likely be the method I pursue to keep the camera self contained and not dependent on any external source of re-streaming to http while maintaining an almost real time stream.

**Reply by endertable** [reply]
Same, prefer leaving out middle man devices and prefer fron the device, if possible.  Mjpeg is great for this as you describe above, fast jog creation

**Reply by virmaior** [reply]
also doing something similar and planning to upload it.  I'm doing it with 5 cameras on the same screen.

looks kind of like:
```
11111 2
11111 3
11111 4
11111 5
(live) (pause)
```
can click on a different camera to make it the bigger one.

Right now, I'm only doing 1 frame/second, but I'm hoping to get some actual streaming hls or something like it for the larger one and make the camera count/urls dynamically enterable.

with 5 cameras, sometimes it fails to load an image from one of them in that second.

**Reply by virmaior** [reply]
using the jpeg solution / js refresh model, i was able to crash at least one of my cameras by setting the refresh interval to 10 ms (not my goal per se) but informative.

**Reply by haderp** [reply]
I have been playing with the ffmpeg version on the camera and have a feed that streams out to mpeg-ts up and viewable on vlc using network stream with udp://@:38080 since I told the camera to stream it to the client ip over udp.  I used:

ffmpeg -r 10  -i /dev/video2 -c:v copy -r 10 -f mpegts udp://x.x.x.x:38080 

My thinking was to then use https://github.com/xqq/mpegts.js to embed it on the camera's http server and give me a stream that I can load from the camera itself.  Not sure yet of the efficacy of my plan as I've not setup the server side, but I thought I might share my idea if it's useful.

**Comment by gtxaspec** [comment]
compiling mjpg-streamer segfaults when run on the camera, i'll try a newer version to see if i can get it to compile properly

using HLS on v4l2rtspserver works for me, although the segfault @virmaior experienced only happened once.  I'll try to adjust the source code to see if I can stabilize it.

**Reply by haderp** [reply]
I was able to setup and run entware alongside wz_mini and using the entware opkg mjpeg-streamer package I was able to get as far as actually getting it to the point of starting to encode the stream, but could never manage to get the right format for the output.  Not sure if that information is of any help, or if using the entware patches could help get the build to run, but figured I'd share my experience running the entware build.  Link to entware mjpeg-streamer patches:

https://github.com/Entware/entware-packages/tree/master/multimedia/mjpg-streamer

**Comment by baudev** [comment]
Is there any updates? I'm also interested in having an HTTP stream (but no HLS neither DASH since I don't want to write to the camera's SD card...)

If not, I'm willing to investigate. What do you recommend: https://github.com/valbok/mjpeg-over-http , https://github.com/pikvm/ustreamer or something else?

From what I saw to do cross compilation I've to use buildroot. So I follow these instructions https://github.com/gtxaspec/wz_mini_hacks/blob/master/documentation/developer_environment.md#buildroot-second and I manually add the desired package, right?
